### PR TITLE
adds support to amf3 parsing

### PR DESCRIPTION
**Pull request by sbijoshj**
Adds a few features for amf3 parsing that we needed. 
Figured it might be worth a PR so you can pull what you like. 
Open to questions and comments!

### CODE REVIEW

File: format/amf3/Tools.hx [hunk 1]
```diff
@@ -56,9 +58,15 @@ class Tools {
 			case cast Array:
 				var o : Array<Dynamic> = o;
 				var a = new Array();
-				for(v in o)
-					a.push(encode(v));
+				for(k in Reflect.fields(o))
```
Comment (ncannasse) [inline format/amf3/Tools.hx]: Why not using normal Array iteration here? Using Reflect.fields will break on most platforms.

File: format/amf3/Tools.hx [hunk 2]
```diff
@@ -87,13 +95,67 @@ class Tools {
...
+					Reflect.setField(s, k, ss);
+				}
+				return s;
+			case AArray(oa):
+				var a:Array<Dynamic> = [];
+				// get both associative array and indexed/dense array values
+				for (af in Reflect.fields(oa))
+				{
+					Reflect.setField(a, af, decodeWithAnonymousStruct(Reflect.field(oa, af)));
+				}
```
Comment (ncannasse) [inline format/amf3/Tools.hx]: Same here, you should not use Reflect.setField/Reflect.fields to iterate/set Arrays.

File: format/amf3/Tools.hx [hunk 3]
```diff
@@ -147,13 +209,25 @@ class Tools {
 		return switch( a ) {
 		case AArray(a):
 			var b = [];
-			for (f in a)
-				b.push(decode(f));
+			for (af in Reflect.fields(a))
+				Reflect.setField(b, af, decode(Reflect.field(a, af)));
```
Comment (ncannasse) [inline format/amf3/Tools.hx]: Another one :)

File: format/amf3/Tools.hx (hunk 1 above)
Comment (sbijoshj) [inline format/amf3/Tools.hx]: I would like to use Haxe's feature to add dynamic fields to emulate the associate array values you get in amf3. I suppose we could change this to return an AssociativeArray or Amf3Array `{ m:Map<String,Dynamic>, a:Array }` instead?

File: format/amf3/Tools.hx (hunk 2 above)
Comment (sbijoshj) [inline format/amf3/Tools.hx]: Here in particular I would like to keep the same due to the legacy code in my project. We can remove this or surround it in some appropriate `#if` perhaps?
Thanks, let me know what you think!

File: format/amf3/Tools.hx (hunk 1 above)
Comment (ncannasse) [inline format/amf3/Tools.hx]: Looking at how AMF does yes I think it requires changing AArray from AArray( a : Array<Value> ) to AArray( a : Array<Value>, ?extra : Map<String,Value> )

File: format/amf3/Tools.hx (hunk 2 above)
Comment (ncannasse) [inline format/amf3/Tools.hx]: I don't think it should be #if, maybe you can have your own version of this function in your project, and keep the original version as-it?

File: format/amf3/Tools.hx (hunk 1 above)
Comment (sbijoshj) [inline format/amf3/Tools.hx]: Ok got it.

File: format/amf3/Tools.hx (hunk 3 above)
Comment (sbijoshj) [inline format/amf3/Tools.hx]: Hmm actually, what do I return here now?

Comment (sbijoshj) [inline format/amf3/Tools.hx]: Is Amf3Array fine?

Comment (sbijoshj) [inline format/amf3/Tools.hx]: Seems to compile - will commit after testing.

**Comment by sbijoshj** [comment]
Updated according to suggestions. Any more I can do to help get this in?

**Comment by sbijoshj** [comment]
@ncannasse we've been using this in production for a while now with no problems, have you had a chance to look at anything else that needs changed to get this merged?

File: format/amf3/Reader.hx [hunk 4]
```diff
@@ -25,81 +25,324 @@
...
+		var fixed = i.readByte() != 0;
+		var a:EitherType<Vector<Value>,Array<Value>>;
+		if( fixed )
+			a = new Vector(len);
+		else
+			a = new Array();
+
+		for(r in 0...len)
+		{
+			a[r] = AInt( i.readInt32() );
```
Comment (sbijoshj) [inline format/amf3/Reader.hx]: For now, `readIntVector` can be used to read both vector-int-type and vector-uint-type but high uint values will overflow and go negative on certain targets. If we ever see an `Input.readUInt32()` we can use that here.

**Pull request merged**